Fix part of #10728: Migrate disallowed python functions regex to custom checker

sjiang647 opened the pull request:
## Overview
1. This PR fixes or fixes part of #10728  .
2. This PR does the following: Removes **only function calls** from `general_purpose_linter`'s `BAD_PATTERNS_PYTHON_REGEXP` and moves them into a custom pylint checker using `visit_call`. **These changes currently do not use an updated `.pylintrc`.**

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

Tests added in `pylint_extensions_test.py`

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

sjiang647 committed:
Remove migrated function calls from general_purpose_linter

sjiang647 committed:
Migrate disallowed function calls to custom checker

sjiang647 commented:
@DubeySandeep PTAL at what I have so far. I currently haven't figured out how to use a new `.pylintrc` config for this checker and its options. I tested what you mentioned in your most recent comment
> Pylint internally fetches the configs defined in the .pylintrc file and if we won't add any config in the .pylintrc file then it will use the default value.

but this wasn't the case when I ran the backend tests.

More specifically, I added the line `disallowed-functions=something,somethingelse` to `.pylintrc` and ran the backend tests; supposedly, the long list of default regexs [here](https://github.com/oppia/oppia/pull/11557/files#diff-900df16033a01355552cd3477b8fa114f0fbb1506fb740207a7ec7a803009aacR1924) should have been overwritten and the tests should have failed, but this was not the case and the tests passed normally.

I'm also not sure how to mock the excluded directories and excluded files in `pylint_extensions_test` for [this exclusion check I wrote](https://github.com/oppia/oppia/pull/11557/files#diff-900df16033a01355552cd3477b8fa114f0fbb1506fb740207a7ec7a803009aacR2054-R2063). Any experience on how to do this?

Thanks so much and apologies for the long questions!

DubeySandeep commented on scripts/linters/pylint_extensions.py:
```diff
@@ -1893,6 +1893,204 @@ def visit_compare(self, node):
+            'remove-disallowed-function-calls',
+            'Please remove disallowed function calls.',
+        ),
+        'C0032': (
```
Update the next id to use the value on top*

sjiang647 replied:
Done

DubeySandeep commented on scripts/linters/pylint_extensions.py:
```diff
@@ -1893,6 +1893,204 @@ def visit_compare(self, node):
+                (
+                    re.compile(r'StringIO\('),
+                    b'python_utils.string_io()',
+                    (python_utils_file, python_utils_test_file,),
```
Instead of excluding the entire file should we use the pragama? (Ditto elsewhere)

sjiang647 replied:
This is definitely a better idea, but keep in mind the pragmas would have to be made per-line since we won't be able to specify which specific function call is disallowed.

DubeySandeep replied:
Yeah, I think that's fine! Do you see any issue with that?

sjiang647 replied:
Not that I'm aware of

DubeySandeep commented on scripts/linters/pylint_extensions.py:
```diff
@@ -1893,6 +1893,204 @@ def visit_compare(self, node):
+        {
+            'default': (
+                (
+                    re.compile(r'datetime.datetime.now\(\)'),
```
why do we need regex here?

sjiang647 replied:
I'm following `general_purpose_linter`'s implementation, but I try to do this without regex as well.

DubeySandeep commented on scripts/linters/pylint_extensions.py:
```diff
@@ -1893,6 +1893,204 @@ def visit_compare(self, node):
+        'C0032': (
+            'Please replace %s with %s.',
+            'replace-disallowed-function-calls',
+            'Please replace disallowed function' +
```
Do we need `+` at the ned?

DubeySandeep commented on scripts/linters/pylint_extensions.py:
```diff
@@ -1893,6 +1893,204 @@ def visit_compare(self, node):
+    python_utils_file = b'python_utils.py'
+    python_utils_test_file = b'python_utils_test.py'
+
+    # pylint: disable=disallowed-function-calls,single-line-pragma
```
Why do we need this pragma here?

DubeySandeep commented on scripts/linters/pylint_extensions_test.py:
```diff
@@ -1,3 +1,4 @@
+
```
Remove unwanted changes!

DubeySandeep commented on scripts/linters/pylint_extensions.py:
```diff
@@ -1893,6 +1893,204 @@ def visit_compare(self, node):
+        Args:
+            node: astroid.nodes.Call. Node to access call content.
+        """
+
```
Remove this empty line*

sjiang647 replied:
Done!

DubeySandeep commented on scripts/linters/pylint_extensions.py:
```diff
@@ -1893,6 +1893,204 @@ def visit_compare(self, node):
+                excluded_dirs) in self.config.disallowed_functions:
+            if self._check_exclusion(curr_path, excluded_files, excluded_dirs):
+                continue
+            if func_rgx.search(called):
```
Instead of regex can we check with the function name directly?

sjiang647 replied:
Done!

DubeySandeep reviewed (changes requested):
@sjiang647, Thanks for the PR, I took a quick pass and it looks great I've left a few general comments, PTAL!

Also, regarding testing configs in test files can you please check how the unity test in pylint are written for the `max-line-length` config? Let me know if you need any help, I'll be very happy to help you! :)

sjiang647 committed:
Remove unwanted changes and manual exclusion checking

sjiang647 committed:
Change checks to be based on function name, revert regex tests to general purpose linter

sjiang647 committed:
Update next id

sjiang647 commented:
@DubeySandeep Addressed your comments and made some updates to make the checker check the function name directly rather than using regex (but had to revert some of `general_purpose_linter`'s regex-specific tests like `'urllib\..*urlencode\('` for this).

For unit testing config files in pylint, I just wanted to make sure we're on the same page before looking into it (this could be a separate issue?). I believe unit tests for Oppia in `pylint_extensions_test.py` don't currently account for `.pylintrc` configs. I was testing around with an existing [options](https://github.com/oppia/oppia/blob/11d5f43dc09cbd8d8e79d9adebd94f158b10aa70/scripts/linters/pylint_extensions.py#L1622-L1629) and [config](https://github.com/oppia/oppia/blob/develop/.pylintrc#L64) (using these existing [unit tests](https://github.com/oppia/oppia/blob/11d5f43dc09cbd8d8e79d9adebd94f158b10aa70/scripts/linters/pylint_extensions_test.py#L2169)).

These unit tests don't use anything from the config, and I'm not sure how else we currently test the config (I noticed when pushing, a message pops up saying `Using config file [path]/.pylintrc`, but I don't know what this message is from. Is our desired behavior to have the pylintrc included in the pylint extension unit tests? Is that why we want to take a look at the pylint repo itself (where `max-line-length` is)?

iamprayush commented:
@DubeySandeep PTAL, thanks!
@sjiang647 just for future reference, make sure that you update the assignee list with the people that you want a review/reply from to make sure that they don't miss it, thanks!

DubeySandeep commented:
> @DubeySandeep Addressed your comments and made some updates to make the checker check the function name directly rather than using regex (but had to revert some of general_purpose_linter's regex-specific tests like 'urllib\..*urlencode\(' for this).

Thanks for the updates, @sjiang647! In general, we/author reply to all the comments (reply "Done" if it's resolved as per the comment's suggestion) and leave it for reviewers to mark resolved (This helps us ensure we are on the same page. :) ) [As documented in the [6th bullet of the getting started wiki page](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#address-review-comments-until-all-reviewers-give-lgtm-looks-good-to-me).]

Can you please reply to all the above comments? :)

> For unit testing config files in pylint, I just wanted to make sure we're on the same page before looking into it (this could be a separate issue?). I believe unit tests for Oppia in pylint_extensions_test.py don't currently account for .pylintrc configs. I was testing around with an existing options and config (using these existing unit tests).

Ah, I see thanks for explaining this in detail. I agree that we should not use the .pylintrc file for testing, I suggest you see how we can add custom config while writing tests (Not sure wither options_providers in pylint test will be helpful). In general I want the pylint_extenstion.py (and corresponding test file) file to be independent of the list of "unallowed functions" basically the check should be a general check and it allows users to add more items through the .pylintrc file. (Note: We don't have to write a test for each disallowed function, we just have to write a test to ensure that the general check structure is working fine.)

Is this^ helpful? (I've sent you an invite on hangouts, feel free to ping me there for a quick discussion. Thanks!)

sjiang647 committed:
Move options to config file. Use custom config in unit tests

sjiang647 committed:
Switch to string-encoding and use ast.literal_eval

sjiang647 committed:
Split up replace and remove disallowed functions into two checkers

sjiang647 commented:
@DubeySandeep Thanks for the tips and review! I'll keep it in mind.

I made changes to support using `.pylintrc` as a config. The way pylint [parses the config](https://github.com/PyCQA/pylint/blob/9a5e1b349fd2521ef8eedaf92b16785c895aaefa/pylint/utils/utils.py#L353-L356) makes it difficult to use tuples as I used for the unit tests in the config file (because the commas in the tuples would cause uneven splitting [eg. (a,b) -> "(a" and "b)"]).

To combat this I decided to just create two options that just take a list of strings that have a 1-to-1 correspondence based on the indices of the values (func name [`disallowed-func-replace`] -> replacement message [`disallowed-func-replacements`]).

I also created a separate checker for functions that should be removed (and not replaced) due to the new structure of replacement options. I wrote a simple test for that, but haven't migrated anything significant into it yet.

What do you think about these changes? PTAL and thanks again!
> I've sent you an invite on hangouts, feel free to ping me there for a quick discussion

Also I couldn't find the google hangouts. Is it via email?

sjiang647 commented:
Updating assignees @DubeySandeep PTAL

DubeySandeep commented:
> I made changes to support using .pylintrc as a config. The way pylint parses the config makes it difficult to use tuples as I used for the unit tests in the config file (because the commas in the tuples would cause uneven splitting [eg. (a,b) -> "(a" and "b)"]).

Good catch! Thanks for letting me know, I'll add this to the "How to write pylint checks doc". I found out that pylint directly uses "," to strip the string. ([ref](https://github.com/PyCQA/pylint/blob/548b38e0508cf58ee9c2fb2e92c0fbb78a1774ef/pylint/utils/utils.py#L353)) **Should we also update the other checker which is using this format?** (maybe in a separate PR?)

> To combat this I decided to just create two options that just take a list of strings that have a 1-to-1 correspondence based on the indices of the values (func name [disallowed-func-replace] -> replacement message [disallowed-func-replacements]).


I would suggest having this pattern: `bad-name | good-name` or `=>` i.,e have our own separator for good-name and bad name. For the function which doesn't have a replacement, we will just have the bad-name. WDYT? (We have to add documentation in the help section of the option.)

Having two options that are linked via index (or are linked in general) for one use case doesn't sound good as it will be hard to manage and users (updating .pylintrc) can easily make mistakes unknowingly which can lead to other developers getting wrong lint messages. what do you think of having our own separator to indicate replacement function?

DubeySandeep commented:
Hi @sjiang647, any update on this PR? Do let me know if you have any other ideas or questions. Thanks!

sjiang647 committed:
Merge from upstream/develop

sjiang647 committed:
Change disallowed functions input to follow format A=>B, remove index-linked inputs

sjiang647 commented:
@DubeySandeep PTAL! 
> I would suggest having this pattern: bad-name | good-name or =>

Updated the checker to use `=>` and changed the corresponding tests.

> Should we also update the other checker which is using this format? (maybe in a separate PR?)

Which other checker are you referencing here? 

Thanks!

sjiang647 committed:
Make replacement messages more consistent, migrate new general_purpose checks to config

sjiang647 committed:
Bypass replacement in python_utils implementations

DubeySandeep commented:
Assigning @sajalasati to take the first pass on the PR!

DubeySandeep commented on .pylintrc:
```diff
@@ -63,6 +63,8 @@ indent-string='    '
 
 allowed-comment-prefixes=int,str,float,bool,v
 
+disallowed-functions=now=>datetime.datetime.utcnow,assertEquals=>self.assertEqual,StringIO=>python_utils.string_io,urlsplit=>python_utils.url_split,urlparse=>python_utils.url_parse,urlunsplit=>python_utils.url_unsplit,parse_qs=>python_utils.parse_query_string,unquote=>python_utils.urllib_unquote,urljoin=>python_utils.url_join,next=>python_utils.NEXT,range=>python_utils.RANGE,round=>python_utils.R ...
```
maybe, `disallowed-functions-with-alternatives` or `replacable-dissalowed-functions`?

Also, for the function which are disallowed and have no alternatives can we use `[bad-builtin](https://github.com/PyCQA/pylint/blob/548b38e0508cf58ee9c2fb2e92c0fbb78a1774ef/tests/extensions/test_bad_builtin.py)` checker?

sjiang647 replied:
Will rename the config. For those without alternatives, I think that's okay, but could we do that on a separate issue/PR?

DubeySandeep replied:
>  I think that's okay, but could we do that on a separate issue/PR

Yep, I agree, can you create an issue to track this task, and let's do it after this PR? :)

DubeySandeep commented on scripts/linters/pylint_extensions.py:
```diff
@@ -1893,6 +1893,81 @@ def visit_compare(self, node):
+    priority = -1
+    msgs = {
+        'C0031': (
+            'Please do not use %s.',
+            'remove-disallowed-function-calls',
+            'Please remove disallowed function calls.',
```
The 1st param should be a user-facing message for the developer so it should be a simple action for them to do (maybe `Please remove %s function call.` [or something around that])

The 3rd param should be a user-facing message for the developer trying to check what this check class does [using pylint --help] (maybe `Disallows usage of black-listed functions.` [or something around that]) 

Ditto for other messages below*

sjiang647 replied:
Done!

DubeySandeep commented on scripts/linters/pylint_extensions.py:
```diff
@@ -1893,6 +1893,81 @@ def visit_compare(self, node):
+                'calls with allowed alternative.'),
+        ),
+    }
+    python_utils_file = b'python_utils.py'
+    python_utils_test_file = b'python_utils_test.py'
```
Do we still need these varaibles?

sjiang647 replied:
Nope. Deleted and done!

sjiang647 replied:
Removed!

DubeySandeep commented on scripts/linters/pylint_extensions.py:
```diff
@@ -1893,6 +1893,81 @@ def visit_compare(self, node):
+        disallowed_entries = self.config.disallowed_functions
+
+        func = node.func
+        for entry in disallowed_entries:
+            splits = [s.strip() for s in entry.split('=>')]
+            assert len(splits) > 0 and len(splits) <= 2
```
Have a separate private function in this class to parse and return the entries (that will make this method more redable and short.

sjiang647 replied:
Done

DubeySandeep commented on scripts/linters/pylint_extensions.py:
```diff
@@ -1893,6 +1893,81 @@ def visit_compare(self, node):
+            splits = [s.strip() for s in entry.split('=>')]
+            assert len(splits) > 0 and len(splits) <= 2
+            func_name = splits[0]
+            if (
+                    (hasattr(func, 'attrname') and func.attrname == func_name)
+                    or (hasattr(func, 'name') and func.name == func_name)):
```
Can you please add a comment above why we need hasattr, func.attrname checks? :)

sjiang647 replied:
Done

DubeySandeep commented on scripts/linters/pylint_extensions.py:
```diff
@@ -1893,6 +1893,81 @@ def visit_compare(self, node):
+                else:
+                    replacement_msg = splits[1]
+                    if (
+                            'python_utils' in node.root().file
+                            and 'python_utils' in replacement_msg):
```
Why do we need this check? (Note this file should be independent of anything in the oppia codebase.)

sjiang647 replied:
Removed!

DubeySandeep commented on scripts/linters/pylint_extensions.py:
```diff
@@ -1893,6 +1893,81 @@ def visit_compare(self, node):
+                            'python_utils' in node.root().file
+                            and 'python_utils' in replacement_msg):
+                        # Allow python_utils to use disallowed function
+                        # in implementing the allowed function.
```
Can we use pragma in that file?

sjiang647 replied:
Done! I hope

DubeySandeep commented on scripts/linters/pylint_extensions_test.py:
```diff
@@ -2915,3 +2915,164 @@ def test_usage_of_is_not_on_none_does_not_add_message(self):
+
+
+class DisallowedFunctionsCheckerTests(unittest.TestCase):
+    """Unit tests for DisallowedFunctionsChecker
+    """
```
```suggestion
    """Unit tests for DisallowedFunctionsChecker."""
```

sjiang647 replied:
Done!

DubeySandeep commented on scripts/linters/pylint_extensions_test.py:
```diff
@@ -2915,3 +2915,164 @@ def test_usage_of_is_not_on_none_does_not_add_message(self):
+
+    def test_disallowed_replacements(self):
+        self.checker_test_object.checker.config.disallowed_functions = [
+            b'now=>datetime.datetime.utcnow',
+            b'assertEquals=>self.assertEqual',
+            b'StringIO=>python_utils.string_io',
+            b'urlsplit=>python_utils.url_split',
+            b'urlparse=>python_utils.url_parse',
+            b'urlunsplit=>python_utils.url_unsplit',
+            b'parse_qs=>python_utils.parse_query_string',
+            b'unquote=>python_utils.urllib_unquote',
+            b'urljoin=>python_utils.url_join',
+            b'next=>python_utils.NEXT',
+            b'range=>python_utils.RANGE',
+            b'round=>python_utils.ROUND',
+            b'str=>python_utils.convert_to_bytes or python_utils.UNICODE',
+            b'zip=>python_utils.ZIP',
+            b'basestring=>python_utils.BASESTRING',
+            b'iteritems=>items',
+            b'itervalues=>values',
+            b'iterkeys=>keys',
+        ]
```
[optional] We don't have to test each of the functions written in the `.pylintrc` (we can assume .pylintrc to be independent). Maybe just have a few cases here?

sjiang647 replied:
Done, reduced to 4 cases.

DubeySandeep reviewed (changes requested):
@sjiang647, this looks great now, I've left a few more comments PTAL!

sjiang647 committed:
Rename config, make messages clearer, reduce tests for disallowed functions, add helper to visit_call, remove general purpose linter tests, and more.

sjiang647 commented:
@DubeySandeep PTAL! Addressed your comments. Also is there a way I can debug test coverage? The failing backend test says for `general_purpose_linter` is missing 532, but I'm not sure what the 532 refers to.

DubeySandeep commented:
@sjiang647, I think you have removed some test which was earlier testing the [line 532 of general_purpose_linter](https://github.com/oppia/oppia/blob/f86d08c771ba43688cf1a52a934cb568bf5918fc/scripts/linters/general_purpose_linter.py#L532). Audit the tests which you have removed and if you won't find any such thing then may add a test for the situation which covers line 532.

DubeySandeep commented on python_utils.py:
```diff
@@ -38,6 +38,7 @@
 import past.utils  # isort:skip  pylint: disable=wrong-import-position, wrong-import-order
 import six  # isort:skip  pylint: disable=wrong-import-position, wrong-import-order
 
+# pylint: disable=single-line-pragma, disallowed-function-calls
```
Should we do it inline wherever needed? to avoid the situation:

```
def new_function_in_python_utils:
    ...
    ...
    ...
   using_and_disallowed_function(...) # Reviewer won't notice this as this isn't using any pragma to avoid usage of disallowed functions.
   ...
    ...
```

sjiang647 replied:
Done

DubeySandeep commented on scripts/linters/general_purpose_linter.py:
```diff
@@ -367,13 +367,6 @@
-        'excluded_dirs': ()
-    },
     {
         'regexp': re.compile(r'ndb\.'),
```
Have a separate DisallowedModulesCheckers check. (I'll create a create issue for this)

DubeySandeep commented on scripts/linters/general_purpose_linter.py:
```diff
@@ -409,27 +402,12 @@
         'excluded_files': (),
         'excluded_dirs': ()
     },
```
Can we add `print=>python_utils.PRINT` for the regex check on line 391?

```
{
        'regexp': re.compile(r'\sprint\('),
        'message': 'Please use python_utils.PRINT().',
        'excluded_files': ('python_utils.py',),
        'excluded_dirs': ()
    }
```

sjiang647 replied:
Done

DubeySandeep commented on scripts/linters/general_purpose_linter.py:
```diff
@@ -409,27 +402,12 @@
         'excluded_files': (),
```
Move this check to the SingleLinePragamChecker.

sjiang647 replied:
Is there something I need to do for this comment, or is this a future plan?

DubeySandeep commented on scripts/linters/general_purpose_linter.py:
```diff
@@ -409,27 +402,12 @@
-        'excluded_dirs': ()
-    },
     {
         'regexp': re.compile(r'with open\(|= open\('),
```
Can we move this to the .pylintrc `open=>python_utils.open_file`?

sjiang647 replied:
Done!

DubeySandeep commented on scripts/linters/general_purpose_linter.py:
```diff
@@ -409,27 +402,12 @@
-        'excluded_dirs': ()
-    },
     {
         'regexp': re.compile(r'urllib\..*quote\('),
```
Move all the urlib related checks to DisallowedModulesCheckers

DubeySandeep commented on scripts/linters/general_purpose_linter.py:
```diff
@@ -514,90 +456,18 @@
         'excluded_files': (),
```
Can we move the input and map function check (line 448) to disallowed-functions-and-replacements?

sjiang647 replied:
Done!

DubeySandeep commented on scripts/linters/pylint_extensions.py:
```diff
@@ -1893,6 +1893,92 @@ def visit_compare(self, node):
+            }
+        ),)
+
+    def _parse_config_entry(self, entry):
```
Maybe change this to _fetch_disallowed_functions_and_replacements? (And also make this function independent of `entry` param as it can independently fetch config through self*)

DubeySandeep replied:
Will it be possible to parse this while initializing the checker and attach this to the object/self? So that we don't have to parse the config n times for n function calls in the module/file*

sjiang647 replied:
I was able to make a change such that it should only fetch once, but was not able to do that at initialization time. Left an explanation in the code.

DubeySandeep replied:
Can we parse the disallowed function list in `open` stage? ([ref](https://github.com/PyCQA/pylint/blob/548b38e0508cf58ee9c2fb2e92c0fbb78a1774ef/pylint/interfaces.py#L57)) [Example: [example1](https://github.com/PyCQA/pylint/blob/60059ae696f745ac3bd8ab7010b63ac8aadb95eb/pylint/checkers/refactoring/refactoring_checker.py#L324), [example2](https://github.com/PyCQA/pylint/blob/f24d765d9fb68309d274c0bd1637da5387cb1f80/pylint/checkers/async.py#L40))

Alternative: Parse the values in `process_modules` method.

sjiang647 replied:
Done with the `open` function!

DubeySandeep commented on scripts/linters/pylint_extensions.py:
```diff
@@ -1893,6 +1893,92 @@ def visit_compare(self, node):
+                just Name.
+
+        Returns:
+            list(str). The parsed function name and an optional replacement.
```
tuple(str, str)*

sjiang647 replied:
This was removed. Done!

DubeySandeep reviewed (changes requested):
@sjiang647, This looks great, I've left a few final comments, PTAL! (Some of the comments are related to creating issues which I'll do once this PR will get merged/approved, feel free to ignore those comments.)

DubeySandeep commented on scripts/linters/general_purpose_linter_test.py:
```diff
@@ -657,60 +556,6 @@ def test_invalid_use_of_object(self):
 
-    def test_invalid_use_of_range(self):
-        linter = general_purpose_linter.GeneralPurposeLinter(
-            [INVALID_RANGE_FILEPATH], FILE_CACHE)
```
Remove `INVALID_RANGE_FILEPATH` file if it's no more required, ditto for other cases.

sjiang647 replied:
Cleaned these variables up as well as the test files associated with them. Done

sjiang647 committed:
Migrate more tests, cleanup old tests, and change custom checker to only read config once.

sjiang647 commented:
@DubeySandeep PTAL. Thanks for the review and bringing up some good points. I've addressed your comments, but I'm still not sure about
> I think you have removed some test which was earlier testing the line 532 of general_purpose_linter

I had audited the tests before and confirmed that each test I removed had a corresponding entry removed from the bad python regex list. Also [line 532](https://github.com/oppia/oppia/blob/f86d08c771ba43688cf1a52a934cb568bf5918fc/scripts/linters/general_purpose_linter.py#L532) there is a single `continue` statement. What do you mean by "covering " this?

Thanks again!

sjiang647 committed:
Merge remote-tracking branch 'upstream/develop' into disallowed-funcs-regex-custom-checker

sajalasati commented:
Hi @sjiang647! To answer your [question](https://github.com/oppia/oppia/pull/11557#issuecomment-760758616), I think None of your tests currently passes the `if` check on line [507](https://github.com/oppia/oppia/blob/3fc25aa99c2c4e0dc45d7e0248daf690d8724231/scripts/linters/general_purpose_linter.py#L507), so the execution never reaches the `continue` statement. This means you have to write a test that executes line [508](https://github.com/oppia/oppia/blob/3fc25aa99c2c4e0dc45d7e0248daf690d8724231/scripts/linters/general_purpose_linter.py#L508). Does that answer your question?

sajalasati commented:
@sjiang647 
Some general notes:
- Don't forget to assign the reviewer and un-assign yourself (as per [this](https://github.com/oppia/oppia/pull/11557#issuecomment-753630101)). If you still want @DubeySandeep to look at this, kindly assign him.
- Note the failing lint checks! (You can see the detailed log and fix it accordingly)
- Before requesting for a re-review, do ensure that you have addressed all the review comments, by replying with a 'done' or asking back a question. (like [this](https://github.com/oppia/oppia/pull/11557#discussion_r557628283) is awaiting reply). Sorry if you are still working on it, but just wanted to bring it to your notice.

Thanks for your PR! Sorry for the delay in reviewing. It looks great to me on a high-level, and I will go ahead to do a detailed review once you resolve the failing tests and address all comments. Does that sound okay @DubeySandeep?

DubeySandeep commented:
> Before requesting for a re-review, do ensure that you have addressed all the review comments, by replying with a 'done' or asking back a question. (like this is awaiting reply). Sorry if you are still working on it, but just wanted to bring it to your notice.

@sajalasati, Sorry for the confusion, some of the above comments need my actions to resolve (i.e, create an issue for them), I'have asked @sjiang647 to leave those comment as is so that I can create an issue once we have covered everything in this PR (as per [this comment](https://github.com/oppia/oppia/pull/11557#pullrequestreview-568524964)). I confirm that @sjiang647 has resolved all of the comments he was expected to resolve! :)

DubeySandeep commented:
@sjiang647, I've left a comment for you over [here](https://github.com/oppia/oppia/pull/11557#discussion_r558479453) PTAL! :)

sajalasati commented:
> @sajalasati, Sorry for the confusion, some of the above comments need my actions to resolve (i.e, create an issue for them), I'have asked @sjiang647 to leave those comment as is so that I can create an issue once we have covered everything in this PR (as per [this comment](https://github.com/oppia/oppia/pull/11557#pullrequestreview-568524964)). I confirm that @sjiang647 has resolved all of the comments he was expected to resolve! :)

@DubeySandeep thanks for clarifying, I had missed that in the thread. Should you still be assigned to the PR in that case? (It will help avoid confusion, oppiabot had removed you automatically)

@sjiang647 I still think it would be a good idea for you, to first try to resolve the failing backend tests and [lint checks](https://github.com/oppia/oppia/pull/11557/checks?check_run_id=1707425135#step:7:317) (shown below too, I think you will need to make some changes in the checker you wrote).
```
  File "scripts/linters/pylint_extensions.py", line 1979, in visit_call
316
    func_key = func.name
317
AttributeError: 'Subscript' object has no attribute 'name'
```

sjiang647 committed:
Fix test coverage, refactor disallowed function checker to use open to populate config

sjiang647 committed:
Merge from upstream/develop

sjiang647 committed:
Fix linting

sjiang647 committed:
Add ignore pragmas to previously excluded tests

sjiang647 committed:
Fix linting for editor_test and jobs_test

sjiang647 committed:
Remove early return

sjiang647 committed:
Fix lint tests

sjiang647 commented:
@sajalasati Appreciate the tips! Will keep them in mind.

@DubeySandeep PTAL

I migrated over the reading of the the config to `open`, and made several small commits to fix lint and backend tests (not sure if it's possible to amend pushed commits without force pushing).

Also to cover the [continue statement for `general_purpose_linter`](https://github.com/oppia/oppia/blob/f86d08c771ba43688cf1a52a934cb568bf5918fc/scripts/linters/general_purpose_linter.py#L532) (which was previously covered in the `iteritems` test), I created a new test file specifically to test out the general purpose linter's pragma. Let me know if there's a better way to do this.

Thanks!

DubeySandeep commented on core/domain/exp_services_test.py:
```diff
@@ -1987,7 +1987,7 @@ def test_export_to_zip_file(self):
         self.assertEqual(zf.namelist(), ['A title.yaml'])
         self.assertEqual(
-            zf.open('A title.yaml').read(), self.SAMPLE_YAML_CONTENT)
+            zf.open('A title.yaml').read(), self.SAMPLE_YAML_CONTENT) # pylint: disable=disallowed-function-calls
```
This should not be disallowed, right? We are only disallowing open(...) function.

sjiang647 replied:
Do you mean we should only be disregarding the python builtin function `open()`, rather than the attribute `zf.open()`? The checker currently disallows attributes with disallowed names, which includes `zf.open()` (rather than just disallowing `open()`).

If so, then there might be some changes required for determining whether we want to disallow attributes or just base functions in how we make the config. One way I can think of is adding some sort of special character in the beginning of the config entry (like `*attr=>new_attr`). If I misinterpreted before and we only want to disallow base functions, then I can change the logic.

DubeySandeep replied:
> Do you mean we should only be disregarding the python builtin function open(), rather than the attribute zf.open()? 

Yes! In general, this is a refactor/restructure PR and we don't expect this PR to changes any old functionality. [Note: Earlier we were only blocking the global open calls.]

Is there any issue with having open=>python_utils.open will only raise the issue if someone uses global open function and something.open=>python_utils.open will only raise the issue if someone uses something.open function? will it be possible to find the module name from the node in the visit_call function? [I'm not sure why we need a special character to represent this?]

sjiang647 replied:
1. That makes sense. However, there's some functionality that we can't exactly capture by simply using a list of exact string replacements, such as the regex [`r'\.put_async\('` that was used in `general_purpose_linter`](https://github.com/oppia/oppia/pull/11557/files#diff-2ec95fb28a886d8e7cf26b726170551779907cbe7dde6bdd0f268076d70f4682L588-L600). What would we do in this case?

2. [**Please correct me if I'm wrong**] I think earlier in this PR you mentioned that I should not use the astroid nodes' [`as_string()` method](http://pylint.pycqa.org/projects/astroid/en/latest/api/base_nodes.html#astroid.node_classes.NodeNG.as_string). I think the complexity of this checker would rise drastically if we need to differentiate between attributes and global calls while strictly using astroid nodes. For example, the config `a.b.c=>d,a.b.c=>e` would require some trie like structure to store (`a->b->c->[d*,e*]` where `*`s represent leaves for lack of better representation). Then, checking in `visit_call` would require searching through the storage while doing type assertions on the underlying nodes (e.g. `b[Name].c[Attribute]()` is different than `a[Name].b[Attribute].c[Attribute]()`).

Apologies if I'm overcomplicating this and for the long PR. I think if we want to have the same functionality as we did in `general_purpose_linter`, it would be best to find a way to have the config use regex and to use the `as_string()` method on the astroid Call node to do the checking. Can you think of any other ways / are there simplifications we can make?

DubeySandeep replied:
> I should not use the astroid nodes' as_string() method.

Yeah, that's true we try to avoid string/regex checks wherever possible. I agree with you that this needs to be done via as string (and we may have to accept function-call-regex from .pylintrc and do some regex check for the as_string text for the cases like put_async.)

I'm fine with using as_string for this checker and if required you can change it to regex function-call pattern in the future (if you are finding it difficult to do in this PR!).

Also, your explanation and questions were very clear, thanks for all the info and findings. 

> . Can you think of any other ways / are there simplifications we can make?

I was thinking of some alternatives but it can be very complicated, I would prefer having as_string in this case!

DubeySandeep commented on core/domain/fs_domain.py:
```diff
@@ -159,7 +159,7 @@ def get(self, filepath):
                 '/%s/%s/%s' % (
                     bucket_name, self._assets_path, filepath))
-            gcs_file = cloudstorage.open(gcs_file_url)
+            gcs_file = cloudstorage.open(gcs_file_url) # pylint: disable=disallowed-function-calls
```
Ditto*

DubeySandeep commented on scripts/linters/test_files/valid_py_ignore_pragma.py:
```diff
@@ -32,13 +34,10 @@ def __init__(self, fake_arg):
         """
-        name = name.iteritems() // # The iteritems() is not allowed to use.
-        yield (name, name)   # disable-bad-pattern-check
+        yield (name, name) # disable-bad-pattern-check
```
Add a newline at the end of the file.

DubeySandeep replied:
@sjiang647, Can you please fix this issue?

sjiang647 replied:
Done

DubeySandeep reviewed (commented):
@sjiang647, This looks good, I've left a few comments, PTAL!

sjiang647 commented:
@DubeySandeep PTAL! left a quick clarification question

DubeySandeep commented:
@sjiang647, I've left a reply [here](https://github.com/oppia/oppia/pull/11557#discussion_r563348422), PTAL! :)

sjiang647 commented:
@DubeySandeep PTAL. Thanks and left a response [here](https://github.com/oppia/oppia/pull/11557#discussion_r563455488).

DubeySandeep commented:
@sjiang647, I've left a reply [here](https://github.com/oppia/oppia/pull/11557#discussion_r566478303), PTAL! :)

sjiang647 committed:
Add regex checker config

sjiang647 committed:
Merge remote-tracking branch 'upstream/develop' into disallowed-funcs-regex-custom-checker

sjiang647 commented:
@DubeySandeep PTAL! Added checker functionality to take in regex from config.

vojtechjelinek commented on .pylintrc:
```diff
@@ -63,6 +63,9 @@ indent-string='    '
 
 allowed-comment-prefixes=int,str,float,bool,v
 
+disallowed-functions-and-replacements-str=now=>datetime.datetime.utcnow,assertEquals=>self.assertEqual,StringIO=>python_utils.string_io,urlsplit=>python_utils.url_split,urlparse=>python_utils.url_parse,urlunsplit=>python_utils.url_unsplit,parse_qs=>python_utils.parse_query_string,unquote=>python_utils.urllib_unquote,urljoin=>python_utils.url_join,next=>python_utils.NEXT,range=>python_utils.RANGE, ...
```
It might be better to add a newline after each `,` ti improve the readability. The newlines then need to be properly cleaned the loaded.

sjiang647 replied:
Done! The config was being trimmed before so I believe that should take care of the cleaning of the newlines.

DubeySandeep commented on core/domain/exp_services_test.py:
```diff
@@ -2073,7 +2073,7 @@ def test_export_to_zip_file_with_assets(self):
         self.assertEqual(
             zf.open('A title.yaml').read(), self.SAMPLE_YAML_CONTENT)
-        self.assertEqual(zf.open('assets/image/abc.png').read(), raw_image)
+        self.assertEqual(zf.open('assets/image/abc.png').read(), raw_image) # pylint: disable=disallowed-function-calls
```
Why do we need this change? (Ditto below)

sjiang647 replied:
Cleaned up the useless pragmas.

DubeySandeep reviewed (commented):
@sjiang647, The changes LGTM, can you please resolve the remain minor comments and we can merge this PR after that?

sjiang647 committed:
Disalled Functions Checker: Add newlines to pylintrc for legibility, remove unused pragmas

sjiang647 committed:
Merge remote-tracking branch 'upstream/develop' into disallowed-funcs-regex-custom-checker

sajalasati commented:
Hi @sjiang647! You don't need to mark the PR as draft to address review comments. After you are done, just un-assign yourself and assign the reviewers.

sjiang647 commented:
Hey @sajalasati appreciate the heads up! I'm drafting the PR while waiting to confirm that the tests pass so that I can make necessary fixes and potentially save time for the reviewers. I also haven't replied to comments yet, and since I'm not a contributor I think I have to use OppiaBot to assign reviewers via comments. Once I confirm that my code is compliant with the checks and I've responded to all comments, I publish the PR for review again just to minimize confusion.

sajalasati commented:
Thanks for explaining your reasoning @sjiang647! 

> I think I have to use OppiaBot to assign reviewers via comments.

Yes, that's correct

> I'm drafting the PR while waiting to confirm that the tests pass so that I can make necessary fixes and potentially save time for the reviewers.

I understand, and it's fine this way. I just wanted to tell you that the general understanding is unless you assign someone to a PR, it understood that you are still working on making some changes. Converting it to a draft might make someone think that some drastic change is needed (like the approach was found to be wrong or related things) or the solution isn't complete yet. So you can keep just yourself assigned if you don't want others to review it yet. I hope you get what I wanted to say now!

sjiang647 commented on .pylintrc:
```diff
@@ -67,6 +67,33 @@ forbidden-imports=
+disallowed-functions-and-replacements-str=now=>datetime.datetime.utcnow,
+assertEquals=>self.assertEqual,
+StringIO=>python_utils.string_io,
+urlparse.urlsplit=>python_utils.url_split,
+urlparse.urlparse=>python_utils.url_parse,
+urlparse.urlunsplit=>python_utils.url_unsplit,
+urlparse.parse_qs=>python_utils.parse_query_string,
+urlparse.urljoin=>python_utils.url_join,
+urllib.unquote=>python_utils.urllib_unquote,
+urllib.urlparse.unquote=>python_utils.urllib_unquote,
```
Updated to use urlparse and urllib modules, since these are the functions whose implementations are being wrapped in `python_utils`

sjiang647 commented:
@DubeySandeep @vojtechjelinek PTAL! Cleaned up unused pragmas and added newlines to pylintrc (not 100% sure how newlines work with pylintrc. I referenced [this](https://stackoverflow.com/questions/47348786/how-to-break-a-long-line-in-pylintrc-file)). Also changed some config entries to specify urllib/urlparse modules inferred from implementation in `python_utils`.

sjiang647 committed:
Disallowed Functions Checker: Add indentations to pylintrc disallowed functions multiline entries

vojtechjelinek commented:
Changes look good. Thanks!

DubeySandeep commented on scripts/linters/pylint_extensions_test.py:
```diff
@@ -3003,6 +3003,200 @@ def test_usage_of_is_not_on_none_does_not_add_message(self):
+                b'example_func',
+                b'a.example_attr',
+            ]
+        self.checker_test_object.checker.open() # pylint: disable=disallowed-function-calls
```
Do we need the pragma here? (Ditto elsewhere)

sjiang647 replied:
Done, meant to remove this earlier sorry.

DubeySandeep reviewed (approved):
@sjiang647 LGTM, just one minor comment PTAL! (I've approved the PR, let's merge it once the last minor comment is resolved!)

sjiang647 committed:
Disallowed Functions Checker: Remove pragmas in pylint test

sjiang647 commented:
@DubeySandeep PTAL. Addressed your comment!

Commit:
Merge remote-tracking branch 'upstream/develop' into HEAD

kevintab95 commented:
Updated this PR with develop to fix flaky e2e tests.
